Project1/memory_manager.cpp

```diff
@@ -0,0 +1,76 @@
+#include <stdlib.h>
+#include <iostream>
+
+#include "memory_manager.h"
+
+MemoryBlock* first_block = nullptr;
+
+MemoryBlock* new_mem_block(size_t size)
+{
+	MemoryBlock* mem = static_cast<MemoryBlock*>(malloc(sizeof(MemoryBlock) + size));
+	if (mem == nullptr) {
+		throw std::bad_alloc();
+	}
+	mem[0].is_allocated = true;
+	mem[0].next_block = nullptr;
+	mem[0].size = size;
+	return mem;
+}
+
+void* allocate(size_t size)
+{
+	if (first_block == nullptr) {
+		first_block = new_mem_block(size);
+		return static_cast<void*>(first_block + 1);
+	}
+	MemoryBlock* prev_block = nullptr;
+	MemoryBlock* block = first_block;
+	while (block != nullptr) {
+		if ((!block->is_allocated) && block->size >= size) {
```

Comment:
I don't think this is a good idea.
It seems like you were trying to optimize your memory manager but this implementation has alot of issues that far outweigh the efficiency plus:
1. If we allocate then deallocate alot of memory, we are left with all the data we've allocated still taken by the process :(
2. If we allocate alot of big buffers then de-allocate them and allocate alot of small buffers, we're gonna end up with huge buffers containing very little information.

In addition, allocating a memory buffer takes O(n) time (where n is the amount of allocations we've already done) that is very slow, please optimize this.